eZ/Publish/Core/Repository/UserService.php

```diff
@@ -583,13 +593,35 @@ public function loadUserByCredentials($login, $password, array $prioritizedLangu
      * @param string $login User login
      * @param string $password User password
      * @param \eZ\Publish\SPI\Persistence\User $spiUser
+     *
+     * @throws \eZ\Publish\Core\Base\Exceptions\BadStateException if the password is not correctly saved, in which case the update is reverted
      */
     private function updatePasswordHash($login, $password, SPIUser $spiUser)
     {
-        if ($spiUser->hashAlgorithm !== $this->settings['hashType']) {
-            $spiUser->passwordHash = $this->createPasswordHash($login, $password, null, $this->settings['hashType']);
-            $spiUser->hashAlgorithm = $this->settings['hashType'];
-            $this->userHandler->update($spiUser);
+        if ($spiUser->hashAlgorithm === $this->settings['hashType']) {
+            return;
+        }
+
+        $spiUser->passwordHash = $this->createPasswordHash($login, $password, null, $this->settings['hashType']);
+        $spiUser->hashAlgorithm = $this->settings['hashType'];
+
+        $this->repository->beginTransaction();
+        $this->userHandler->update($spiUser);
+        $reloadedSpiUser = $this->userHandler->load($spiUser->id);
+
+        if ($reloadedSpiUser->passwordHash === $spiUser->passwordHash) {
+            $this->repository->commit();
+        } else {
+            // Password hash was not correctly saved, possible cause: EZP-28692
+            $this->repository->rollback();
+            if (isset($this->logger)) {
+                $this->logger->error('Password hash could not be updated. Please verify that your database schema is up to date.');
```

Comment:
As suggested on slack, please change this to critical so it will be logged also on prod in case people are testing with prod on local dev environment. And if this is in prod, it really is critical anyway.